### PR TITLE
Add Meterpreter sanity tests to CI

### PR DESCRIPTION
**Pull request by adfoster-r7**
Revival of https://github.com/rapid7/metasploit-framework/pull/16357

Adds a larger suite of Meterpreter sanity tests to Github Actions

image

We also use Allure to generate an interactive test report:

image

Each generated test has the reproduction steps, and console output available for debugging purposes:

image

Under the hood the test suite spins up msfconsole as a child process and interacts with it via stdin/stdout. The main work flow is generating a payload, creating a listener, executing the payload, validating the stdout of the module. This approach was chosen over pure unit tests to help replicate more closely how a user would interact with msfconsole, and to catch semantic differences, such as autoloading disparity, between the existing test suite versus booting msfconsole directly

## Verification

- Review the code
- Ensure the acceptance tests can be run locally

### CODE REVIEW

File: .github/workflows/acceptance.yml [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+      fail-fast: false
+      matrix:
+        os:
+          - macos-11
+          - windows-2019
+          - ubuntu-20.04
+        ruby:
+          - 3.0.2
+        meterpreter:
+          # Python
```
Comment (jvoisin) [inline .github/workflows/acceptance.yml:58]: Isn't Python2 still supported?

File: .github/workflows/acceptance.yml [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+      RAILS_ENV: test
+      HOST_RUNNER_IMAGE: ${{ matrix.os }}
+      METERPRETER: ${{ matrix.meterpreter.name }}
+      METERPRETER_RUNTIME_VERSION: ${{ matrix.meterpreter.runtime_version }}
+
+    name: ${{ matrix.meterpreter.name }} ${{ matrix.meterpreter.runtime_version }} ${{ matrix.os }}
+    steps:
+      - name: Install system dependencies (Linux)
+        if: runner.os == 'Linux'
+        run: sudo apt-get install libpcap-dev graphviz
```
Comment (jvoisin) [inline .github/workflows/acceptance.yml:92]:
```suggestion
        run: sudo apt-get install -y --no-install-recommends libpcap-dev graphviz
```

File: .github/workflows/acceptance.yml (hunk 1 above)
Comment (adfoster-r7) [reply .github/workflows/acceptance.yml:58]: It's no longer possible to run Python 2.7 natively on Github actions without Docker - https://github.com/actions/setup-python/issues/672 I can revisit this decision again in the future, but we won't get Windows/OSX Python 2.7 testing

For visibility - over in the metasploit-payloads repo I introduced a unit test suite for the Python Meterpreter which runs multiple Python versions as well https://github.com/rapid7/metasploit-payloads/pull/651 It'd be great to add something similar for the PHP payload, but I don't believe there's any native test suite runners that come bundled with PHP like Python's `unittest`

Comment (jvoisin) [reply .github/workflows/acceptance.yml:58]: For PHP, there is [PHPUnit]( https://phpunit.de/ ) (to test PHP code behaviour) and/or [PHP's native testsuite]( https://qa.php.net/running-tests.php ) (to test the interpreter behaviour).

Comment (adfoster-r7) [reply .github/workflows/acceptance.yml:58]: Just created a separate issue for adding a basic PHP Meterpreter test suite https://github.com/rapid7/metasploit-payloads/issues/668 :+1:

For this PR though, there won't be Python 2.7 integration tests running natively on osx/windows/linux due to it being dropped by Github

File: .github/workflows/acceptance.yml [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        with:
+          distribution: temurin
+          java-version: ${{ matrix.meterpreter.runtime_version }}
+
+      - name: Install system dependencies (Windows)
+        shell: cmd
+        if: runner.os == 'Windows'
+        run: |
+          REM pcap dependencies
+          powershell -Command "[System.Net.ServicePointManager]::ServerCertificateValidationCallback = {$true} ; [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; (New-Object System.Net.WebClient).DownloadFile('https://www.winpcap.org/install/bin/WpdPack_4_1_2.zip', 'C:\Windows\Temp\WpdPack_4_1_2.zip')"
```
Comment (dwelch-r7) [inline .github/workflows/acceptance.yml:117]: Just out of curiosity what's all this needed for? and any chance that version will disappear and break the job?

File: lib/msf/core/post/windows/registry.rb [hunk 4]
```diff
@@ -103,7 +102,7 @@ def registry_hive_lookup(hive)
   # Load a hive file
   #
   def registry_loadkey(key, file)
-    if session_has_registry_ext
+    if session.commands.include?(Rex::Post::Meterpreter::Extensions::Stdapi::COMMAND_ID_STDAPI_REGISTRY_LOAD_KEY)
```
Comment (dwelch-r7) [inline lib/msf/core/post/windows/registry.rb:105]: Are these changes related to the tests?

Comment (adfoster-r7) [reply lib/msf/core/post/windows/registry.rb:105]: Yip

File: .github/workflows/acceptance.yml (hunk 3 above)
Comment (adfoster-r7) [reply .github/workflows/acceptance.yml:117]: The pcaprub gem dependency needs it to compile during a bundle on windows

**Comment by dwelch-r7** [comment]
# Release Notes

Adds a larger suite of Meterpreter sanity tests to Github Actions
